x/evm/statedb/statedb.go

```diff
@@ -0,0 +1,481 @@
+package statedb
+
+import (
+	"fmt"
+	"math/big"
+	"sort"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/ethereum/go-ethereum/common"
+	ethtypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/core/vm"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+type revision struct {
+	id           int
+	journalIndex int
+}
+
+var _ vm.StateDB = &StateDB{}
+
+// StateDB structs within the ethereum protocol are used to store anything
+// within the merkle trie. StateDBs take care of caching and storing
+// nested states. It's the general query interface to retrieve:
+// * Contracts
+// * Accounts
+type StateDB struct {
+	keeper Keeper
+	ctx    sdk.Context
+	dbErr  error
+
+	// Journal of state modifications. This is the backbone of
+	// Snapshot and RevertToSnapshot.
+	journal        *journal
+	validRevisions []revision
+	nextRevisionID int
+
+	stateObjects map[common.Address]*stateObject
+
+	txConfig TxConfig
+
+	// The refund counter, also used by state transitioning.
+	refund uint64
+
+	logs []*ethtypes.Log
+
+	// Per-transaction access list
+	accessList *accessList
+}
+
+// New creates a new state from a given trie.
+func New(ctx sdk.Context, keeper Keeper, txConfig TxConfig) *StateDB {
+	return &StateDB{
+		keeper:       keeper,
+		ctx:          ctx,
+		stateObjects: make(map[common.Address]*stateObject),
+		journal:      newJournal(),
+		accessList:   newAccessList(),
+
+		txConfig: txConfig,
+	}
+}
+
+// Keeper returns the underlying `Keeper`
+func (s *StateDB) Keeper() Keeper {
+	return s.keeper
+}
+
+// Context returns the embedded `sdk.Context`
+func (s *StateDB) Context() sdk.Context {
+	return s.ctx
+}
+
+// setError remembers the first non-nil error it is called with.
+func (s *StateDB) setError(err error) {
+	if s.dbErr == nil {
+		s.dbErr = err
+	}
+}
+
+func (s *StateDB) Error() error {
+	return s.dbErr
+}
+
+func (s *StateDB) AddLog(log *ethtypes.Log) {
+	s.journal.append(addLogChange{})
+
+	log.TxHash = s.txConfig.TxHash
+	log.BlockHash = s.txConfig.BlockHash
+	log.TxIndex = s.txConfig.TxIndex
+	log.Index = s.txConfig.LogIndex + uint(len(s.logs))
+	s.logs = append(s.logs, log)
+}
+
+func (s *StateDB) Logs() []*ethtypes.Log {
+	return s.logs
+}
+
+// AddRefund adds gas to the refund counter
+func (s *StateDB) AddRefund(gas uint64) {
+	s.journal.append(refundChange{prev: s.refund})
+	s.refund += gas
+}
+
+// SubRefund removes gas from the refund counter.
+// This method will panic if the refund counter goes below zero
+func (s *StateDB) SubRefund(gas uint64) {
+	s.journal.append(refundChange{prev: s.refund})
+	if gas > s.refund {
+		panic(fmt.Sprintf("Refund counter below zero (gas: %d > refund: %d)", gas, s.refund))
```

Comment:
this shouldn't happen or could it?

Reply:
I think the invariant should be guaranteed by EVM implementation.

Reply:
hmm should we set and log the error instead?

Reply:
I guess panic is ok here since it only happens if a serious bug happens inside evm?